tests/pyfunc/test_pyfunc_schema_enforcement.py

```diff
@@ -914,3 +919,620 @@ def test_schema_enforcement_for_list_inputs():
+    with pytest.raises(MlflowException, match=error_msg):
+        _enforce_params_schema({"str_param": np.datetime64("2023-06-06")}, test_schema)
+
+    error_msg = r"Incompatible types for param 'byte_param'"
+    with pytest.raises(MlflowException, match=error_msg):
+        _enforce_params_schema({"byte_param": np.float32(1)}, test_schema)
+
+    # With Array
+    with pytest.raises(MlflowException, match="Incompatible types for param 'float_array'"):
+        _enforce_params_schema(
+            {"float_array": [np.float32(1), np.float32(2), np.float64(3)]}, schema
+        )
+
+    # Add default values if the parameter is not provided
+    test_parameters = {"a": "str_a"}
+    test_schema = ParamSchema(
+        [ParamSpec("a", DataType.string, ""), ParamSpec("b", DataType.long, 1)]
+    )
+    updated_parameters = {"b": 1}
+    updated_parameters.update(test_parameters)
+    assert _enforce_params_schema(test_parameters, test_schema) == updated_parameters
+
+    # Ignore values not specified in ParamSchema and log warning
+    test_parameters = {"a": "str_a", "invalid_param": "value"}
+    test_schema = ParamSchema([ParamSpec("a", DataType.string, "")])
+    with mock.patch("mlflow.models.utils._logger.warning") as mock_warning:
+        assert _enforce_params_schema(test_parameters, test_schema) == {"a": "str_a"}
+        mock_warning.assert_called_once_with(
+            "Unrecognized params ['invalid_param'] are ignored for inference. "
+            "Supported params are: {'a'}. "
+            "To enable them, please add corresponding schema in ModelSignature."
+        )
+
+    # Converting parameters keys to string if it is not
+    test_parameters = {1: 1.0}
+    test_schema = ParamSchema([ParamSpec("1", DataType.double, 1.0)])
+    assert _enforce_params_schema(test_parameters, test_schema) == {"1": 1.0}
+
+    # Raise error if parameters is not dictionary
+    with pytest.raises(MlflowException, match=r"Parameters must be a dictionary. Got type 'int'."):
+        _enforce_params_schema(100, test_schema)
+
+    # Raise error if invalid parameters are passed
+    test_parameters = {"a": True, "b": (1, 2), "c": b"test"}
+    test_schema = ParamSchema(
+        [
+            ParamSpec("a", DataType.boolean, False),
+            ParamSpec("b", DataType.string, [], (-1,)),
+            ParamSpec("c", DataType.string, ""),
+        ]
+    )
+    with pytest.raises(
+        MlflowException,
+        match=re.escape(
+            "Value must be a 1D array with shape (-1,) for param 'b': string "
+            "(default: []) (shape: (-1,)), received tuple"
+        ),
+    ):
+        _enforce_params_schema(test_parameters, test_schema)
+    # Raise error for non-1D array
+    with pytest.raises(MlflowException, match=r"received list with ndim 2"):
+        _enforce_params_schema(
+            {"a": [[1, 2], [3, 4]]}, ParamSchema([ParamSpec("a", DataType.long, [], (-1,))])
+        )
+
+
+def test_param_spec():
```

Comment:
Same comment on this test: https://github.com/mlflow/mlflow/pull/8563/files#r1252807625